File: src/nim/bedcov.nim
import std/algorithm

type
  Interval*[S,T] = tuple[st, en: S, data: T, max: S]

proc intvcmp[S,T](x, y: Interval[S,T]): int =
  cmp(x.st, y.st)

proc index*[S,T](a: var seq[Interval[S,T]]): int {.discardable.} =
  if a.len == 0: return 0
  var is_srt = true
  for i in 1..<a.len:
    if a[i-1].st > a[i].st:
      is_srt = false; break
  if not is_srt: a.sort(intvcmp)
  var last_i: int
  var last: S
  for i in countup(0, a.len-1, 2): # leaves (i.e. at level 0)
    (last_i, last, a[i].max) = (i, a[i].en, a[i].en)
  var k = 1
  while 1 shl k <= a.len: # process internal nodes in the bottom-up order
    let x = 1 shl (k - 1)
    let i0 = (x shl 1) - 1 # the first node at level k
    let step = x shl 2
    for i in countup(i0, a.len - 1, step): # traverse nodes at level k
      let el = a[i - x].max  # max value of the left child
      let er = if i + x < a.len: a[i + x].max else: last # of the right child
      var e = a[i].en
      if e < el: e = el
      if e < er: e = er
      a[i].max = e
    # point last_i to the parent of the original last_i
    last_i = if ((last_i shr k) and 1) != 0: last_i - x else: last_i + x
    if last_i < a.len and a[last_i].max > last: # update last accordingly
      last = a[last_i].max
    k += 1
  return k - 1

iterator overlap*[S,T](a: seq[Interval[S,T]], st: S, en: S): Interval[S,T] {.noSideEffect.} =
  var h: int = 0
  while 1 shl h <= a.len: h += 1
  h -= 1 # h is the height of the tree
  var stack: array[64, tuple[k, x, w:int]] # 64 is the max possible tree height
  var t: int = 0
  stack[t] = (h, (1 shl h) - 1, 0); t += 1 # push the root
  while t > 0: # the following guarantees sorted "yield"
    t -= 1
    let (k, x, w) = stack[t] # pop from the stack
    if k <= 3: # in a small subtree, traverse everything
      let i0 = (x shr k) shl k
      var i1 = i0 + (1 shl (k + 1)) - 1
      if i1 >= a.len: i1 = a.len
      for i in countup(i0, i1 - 1):
        if a[i].st >= en: break  # out of range; no need to proceed
        if st < a[i].en: yield a[i] # overlap! yield
    elif w == 0: # the left child not processed
      let y = x - (1 shl (k - 1)) # the left child of z.x; y may >=a.len
      stack[t] = (k, x, 1); t += 1
      if y >= a.len or a[y].max > st:
        stack[t] = (k-1, y, 0); t += 1 # add left child
    elif x < a.len and a[x].st < en: # need to push the right child
      if st < a[x].en: yield a[x] # test if x overlaps the query
      stack[t] = (k - 1, x + (1 shl (k - 1)), 0); t += 1

type
  IntervalList64 = seq[Interval[int64,int64]]

type
  Splitmix32 = object
    x*: uint32

proc splitmix32(rng: var Splitmix32): uint32 =
  rng.x = rng.x + 0x9e3779b9'u32
  var z = rng.x
  z = (z xor (z shr 16)) * 0x21f0aaad'u32
  z = (z xor (z shr 15)) * 0x735a2d97'u32
  return z xor (z shr 15)

proc gen_intv(n: int64, rng: var Splitmix32, bit_st: int64, bit_len: int64): IntervalList64 =
  result = @[]
  let mask_st  = (1'i64 shl bit_st)  - 1
  let mask_len = (1'i64 shl bit_len) - 1
  for i in 0..<n:
    let st: int64 = int64(splitmix32(rng)) and mask_st
    let en: int64 = st + (int64(splitmix32(rng)) and mask_len)
    result.add((st, en, int64(i), 0'i64))

var rng = Splitmix32(x:11)
var n = 1000000'i64
var bit_st = 28
var bit_len = 14
var a1 = gen_intv(n, rng, bit_st, bit_len)
var a2 = gen_intv(n, rng, bit_st, bit_len)
a1.index()
var tot_cov: int64 = 0
for i in 0..<n:
  let st0 = a2[i].st
  let en0 = a2[i].en
  var cov_st, cov_en, cov, cnt: int64
  for x in a1.overlap(st0, en0):
    let st1 = if x.st > st0: x.st else: st0
    let en1 = if x.en < en0: x.en else: en0
    if st1 > cov_en:
      cov += cov_en - cov_st
      (cov_st, cov_en) = (st1, en1)
    else:
      cov_en = if cov_en > en1: cov_en else: en1
  cov += cov_en - cov_st
  tot_cov += cov
echo tot_cov

File: src/julia/bedcov.jl
struct Interval{S,T}
	data::T
	st::S
	en::S
	max::S
end

function it_index!(a::Memory{Interval{S,T}}) where {S,T}
	sort!(a, by = x -> x.st)
	last_i = 1
	last::S = 0
	@inbounds for i = 1:2:length(a)
		last, last_i = a[i].en, i;
		a[i] = Interval{S,T}(a[i].data, a[i].st, a[i].en, a[i].en)
	end
	k::Int = 1
	@inbounds while 1<<k <= length(a)
		i0, step = (1<<k), 1<<(k+1)
		for i = i0:step:length(a)
			x = 1<<(k-1)
			max_end = max(a[i].en, a[i-x].max)
			e = if (i + x <= length(a)) a[i+x].max else last end
			max_end = max(max_end, e)
			a[i] = Interval{S,T}(a[i].data, a[i].st, a[i].en, max_end)
		end
		last_i = if ((last_i>>k&1) != 0) last_i - (1<<(k-1)) else last_i + (1<<(k-1)) end
		if (last_i <= length(a)) last = max(last, a[last_i].max) end
		k += 1
	end
end

function it_overlap!(
	a::Memory{Interval{S,T}},
	stack::Memory{NTuple{3, Int}},
	st::S,
	en::S,
	b::Vector{Interval{S,T}}
) where {S,T}
	empty!(b)
	h = 0
	while (1<<h <= length(a)) h += 1 end
	h -= 1
	@inbounds stack[1] = ((1<<h), h, 0)
	stack_len = 1
	@inbounds while stack_len > 0
		x, h, w = stack[stack_len]
		stack_len -= 1
		if h <= 3
			i0 = ((x-1) >> h << h) + 1
			i1 = i0 + (1 << (h+1)) - 2
			i1 = min(i1, length(a))
			i = i0
			while i <= i1 && a[i].st < en
				if (st < a[i].en) push!(b, a[i]) end
				i += 1
			end
		elseif w == 0
			stack[stack_len += 1] = (x, h, 1)
			y = x - (1<<(h-1))
			if y > length(a) || a[y].max > st
				stack[stack_len += 1] = (y, h - 1, 0)
			end
		elseif x <= length(a) && a[x].st < en
			if (st < a[x].en) push!(b, a[x]) end
			stack[stack_len += 1] = (x + (1<<(h-1)), h - 1, 0)
		end
	end
end

mutable struct Splitmix32
	x::UInt32
end

function splitmix32!(rng::Splitmix32)::UInt32
	rng.x += 0x9e3779b9
	z = rng.x
	z = xor(z, z >> 16) * 0x21f0aaad
	z = xor(z, z >> 15) * 0x735a2d97
	z = xor(z, z >> 15)
	return z
end

function gen_intv(n, x::Splitmix32, bit_st, bit_len)
	mask_st  = (1<<bit_st)  - 1
	mask_len = (1<<bit_len) - 1
	a = Memory{Interval{Int64,Int64}}(undef, n)
	for i = 1:n
		s = splitmix32!(x)
		l = splitmix32!(x)
		st = s & mask_st
		en = st + (l & mask_len)
		a[i] = Interval{Int64,Int64}(i, st, en, 0)
	end
	return a, x
end

function main(args)
	bit_st, bit_len = 28, 14
	x = Splitmix32(11)
	n = 1000000
	a1, x  = gen_intv(n, x, bit_st, bit_len)
	a2, x  = gen_intv(n, x, bit_st, bit_len)
	it_index!(a1)
	tot_cov = 0
	b = Vector{Interval{Int64,Int64}}()
	stack = Memory{NTuple{3, Int}}(undef, 64)
	for k = 1:n
		st0, en0 = a2[k].st, a2[k].en
		it_overlap!(a1, stack, st0, en0, b)
		cov_st, cov_en, cov = 0, 0, 0
		for i = 1:length(b)
			st1 = max(b[i].st, st0)
			en1 = min(b[i].en, en0)
			if st1 > cov_en
				cov += cov_en - cov_st
				cov_st, cov_en = st1, en1
			else
				cov_en = max(cov_en, en1)
			end
		end
		cov += cov_en - cov_st
		tot_cov += cov
	end
	println(tot_cov)
end

main(ARGS)
